articles/securing-your-python-application-from-hackers/index.md

```diff
@@ -0,0 +1,174 @@
+4.0
+```
+
+#### Input()
+You can find this vulnerability only in [Python 2](https://www.python.org/downloads/release/python-272/) and you can't find it in [Python 3](https://www.python.org/downloads/).
+
+This function in Python 2 takes the value type, as it is without changing its type. 
+```python
+secret = 5
+number = input("Expect the secret number")
+if number==secret:
+    print("YES")
+else:
+    print("NO")
+```
+
+Now we have a secret number, and we want the user to expect it. If we pass `5` then the result will be `YES` and otherwise it will be `NO`, but if we pass `secret` then the result will be `YES` because the Python conditional now is:
+```python
+if secret==secret: // This will be true
+```
+
+And just like that, you can pass any password leading to authentication bypass. Also, there are a lot of other ways to exploit this function.
+
+To fix this issue, you just need to use `raw_input()` instead of `input()` while you are using Python 2 because it is converting any input to a string. If you are using Python 3 then don't worry about it while it is doing just the same as `raw_input()`.
+
+### String formatting exploiting
+This can be a very dangerous method because it leads the hackers to authentication bypass and to get access to sensitive information. So how does this really work?
+
+Let's build a simple app, assume that *CONFIG* contain sensitive information like an *API key*:
+```python
+CONFIG = {
+    "KEY" : "SECRET_KEY"
+}
+class info:
+    def __init__(self, first, last):
+        self.first = first
+        self.last = last
+
+def getName(some_str, peopleinfo_obj):
+    return some_str.format(peopleinfo_obj = peopleinfo_obj)
+
+people = info('Ahmad', 'Mardeni')
+
+stri = input()
+getName(stri, peopleinfo_obj = people)
+```
+
+Now if the user passes the following as input:
+```python
+Name_is_{peopleinfo_obj.first}_{peopleinfo_obj.last}
+```
+
+Then the output will be:
+```python
+'Name_is_Ahmad_Mardeni'
+```
+
+But if he passes the following as input:
+```python
+{peopleinfo_obj.__init__.__globals__[CONFIG][KEY]}
+```
+
+Then our sensitive information will be compromised and the output will be:
+```python
+'SECRET_KEY'
+```
+
+That happened because the string formatting function has the ability to access attributes objects. And the question now, is it still good to use `str.format()`? Yes, but you have to be aware that it becomes vulnerable when it is used over user-controlled strings.
+
+### Conclusion 
+We learned about the dangerous functions `eval()`, `exec()` ,and `input()`, how hackers can use them to hack your Python code, and finally, how to protect yourself against them. But anyway, you have to pay an attention to vulnerabilities besides the language-specific ones, like [XSS](https://owasp.org/www-community/attacks/xss/), [SQL injection](https://www.w3schools.com/sql/sql_injection.asp), etc...
+
+Till next time, happy coding Pythonistas!
+
+![](https://media.giphy.com/media/LmNwrBhejkK9EFP504/giphy.gif)

```

Comment:
Please add alt text to the image, and include the image in the repo instead of using a link.